drivers/i2c/i2c_litex_litei2c.c

```diff
@@ -0,0 +1,374 @@
+	uint32_t phy_speed_mode_addr;
+	uint32_t master_active_addr;
+	uint32_t master_settings_addr;
+	uint32_t master_addr_addr;
+	uint32_t master_rxtx_addr;
+	uint32_t master_status_addr;
+	uint32_t bitrate;
+};
+
+struct i2c_litex_litei2c_data {
+	struct k_mutex mutex;
+};
+
+static int i2c_litex_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct i2c_litex_litei2c_config *config = dev->config;
+	struct i2c_litex_litei2c_data *data = dev->data;
+
+	if (I2C_ADDR_10_BITS & dev_config) {
+		return -ENOTSUP;
+	}
+
+	if (!(I2C_MODE_CONTROLLER & dev_config)) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	/* Setup speed to use */
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		litex_write8(0, config->phy_speed_mode_addr);
+		break;
+	case I2C_SPEED_FAST:
+		litex_write8(1, config->phy_speed_mode_addr);
+		break;
+	case I2C_SPEED_FAST_PLUS:
+		litex_write8(2, config->phy_speed_mode_addr);
+		break;
+	default:
+		k_mutex_unlock(&data->mutex);
+		return -ENOTSUP;
+	}
+
+	k_mutex_unlock(&data->mutex);
+
+	return 0;
+}
+
+static int i2c_litex_get_config(const struct device *dev, uint32_t *config)
+{
+	const struct i2c_litex_litei2c_config *dev_config = dev->config;
+
+	*config = I2C_MODE_CONTROLLER;
+
+	switch (litex_read8(dev_config->phy_speed_mode_addr)) {
+	case 0:
+		*config |= I2C_SPEED_SET(I2C_SPEED_STANDARD);
+		break;
+	case 1:
+		*config |= I2C_SPEED_SET(I2C_SPEED_FAST);
+		break;
+	case 2:
+		*config |= I2C_SPEED_SET(I2C_SPEED_FAST_PLUS);
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static int i2c_litex_write_settings(const struct device *dev, uint8_t len_tx, uint8_t len_rx,
+				    bool recover)
+{
+	const struct i2c_litex_litei2c_config *config = dev->config;
+
+	uint32_t settings = len_tx | (len_rx << 8) | (recover << 16);
+
+	litex_write32(settings, config->master_settings_addr);
+
+	return 0;
+}
+
+static int i2c_litex_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			      uint16_t addr)
+{
+	const struct i2c_litex_litei2c_config *config = dev->config;
+	struct i2c_litex_litei2c_data *data = dev->data;
+	uint32_t len_tx_buf = 0;
+	uint32_t len_rx_buf = 0;
+	uint8_t len_tx = 0;
+	uint8_t len_rx = 0;
+
+	uint8_t *tx_buf_ptr;
+	uint8_t *rx_buf_ptr;
+
+	uint32_t tx_buf;
+	uint32_t rx_buf;
+
+	uint32_t tx_j = 0;
+	uint32_t rx_j = 0;
+
+	uint8_t *burst_buf;
+
+	int ret = 0;
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	litex_write8(1, config->master_active_addr);
+
+	LOG_DBG("addr: 0x%x", addr);
+	litex_write8((uint8_t)addr, config->master_addr_addr);
+
+	for (uint8_t i = 0; i < num_msgs; i++) {
+		if (msgs[i].flags & I2C_MSG_READ) {
+			len_tx_buf = 0;
+			len_rx_buf = msgs[i].len;
+			rx_buf_ptr = msgs[i].buf;
+			tx_buf_ptr = NULL;
+		} else {
+			len_tx_buf = msgs[i].len;
+			tx_buf_ptr = msgs[i].buf;
+			/* This is only used to be able to do burst writes */
+			if (!(msgs[i].flags & I2C_MSG_STOP) && (i + 1 < num_msgs) &&
```

Comment:
i2c_burst_write is entirely optional and should be viewed as unportable because it needs this sort of message merging.

Drop support for i2c_burst_write instead and the k_malloc/k_free along with it. We shouldn't be doing any allocating on the heap in an i2c transfer. If the hardware doesn't support a feature, that's ok! -ENOSUP it.

Reply:
removed that part